{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os, h5py\n",
    "import pandas as pd\n",
    "# from collections import OrderedDict\n",
    "import matplotlib.pyplot as plt\n",
    "from pyliftover import LiftOver\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta2list(fasta_file):\n",
    "    fasta_coords = []\n",
    "    seqs = []\n",
    "    header = ''\n",
    "\n",
    "    for line in open(fasta_file):\n",
    "        if line[0] == '>':\n",
    "            #header = line.split()[0][1:]\n",
    "            fasta_coords.append(line[1:].rstrip())\n",
    "        else:\n",
    "            s = line.rstrip()\n",
    "            s = s.upper()\n",
    "            seqs.append(s)\n",
    "\n",
    "    return fasta_coords, seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dna_one_hot(seq):\n",
    "\n",
    "    seq_len = len(seq)\n",
    "    seq_start = 0\n",
    "    seq = seq.upper()\n",
    "\n",
    "    # map nt's to a matrix len(seq)x4 of 0's and 1's.\n",
    "    seq_code = np.zeros((seq_len, 4), dtype='float16')\n",
    "\n",
    "    for i in range(seq_len):\n",
    "        if i >= seq_start and i - seq_start < len(seq):\n",
    "            nt = seq[i - seq_start]\n",
    "            if nt == 'A':\n",
    "                seq_code[i, 0] = 1\n",
    "            elif nt == 'C':\n",
    "                seq_code[i, 1] = 1\n",
    "            elif nt == 'G':\n",
    "                seq_code[i, 2] = 1\n",
    "            elif nt == 'T':\n",
    "                seq_code[i, 3] = 1\n",
    "            else:\n",
    "                seq_code[i,:] = 0.25\n",
    "\n",
    "    return seq_code\n",
    "\n",
    "def enforce_const_range(site, window):\n",
    "#     site -= 1 \n",
    "    half_window = np.round(window/2).astype(int)\n",
    "    start = site - half_window\n",
    "    end = site + half_window\n",
    "    return start, end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expand_range(bedfile, output_filename, window=3072):\n",
    "    df = pd.read_csv(bedfile, sep='\\t', header=None, index_col=None)\n",
    "    start, end = enforce_const_range(df.iloc[:,1].astype(int), window)\n",
    "    df_expanded = df.copy()\n",
    "    df_expanded.iloc[:,1] = start.values\n",
    "    df_expanded.iloc[:,2] = end.values\n",
    "    df_nonneg = df_expanded[df_expanded.iloc[:,1]>0]\n",
    "    df_nonneg = df_nonneg.reset_index(drop=True)\n",
    "    df_nonneg.to_csv(output_filename, header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_accessible_regions(basset_samplefile='/mnt/906427d6-fddf-41bf-9ec6-c3d0c37e766f/amber/ATAC/basset_sample_file.tsv',\n",
    "                              output_accessible_regions='combined_atac.bed'):\n",
    "    bed_paths = pd.read_csv(basset_samplefile, sep='\\t', header=None)[1].values\n",
    "    combined_csv = pd.concat([pd.read_csv(f, sep='\\t', header=None) for f in bed_paths ])\n",
    "    combined_csv.to_csv(output_accessible_regions, sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_hg19_to_hg38(bedfile, chain_file='/home/amber/QuantPred/datasets/VCF/hg19ToHg38.over.chain'):\n",
    "    lo = LiftOver(chain_file)\n",
    "    dataset = pd.read_csv(bedfile, header=None, sep='\\t')\n",
    "    hg38_coords = []\n",
    "    for i, (chrom, start) in dataset.iloc[:,:2].iterrows():    \n",
    "        result = lo.convert_coordinate(chrom, start)\n",
    "        if len(result) == 1:\n",
    "            hg38_coords.append(result[0][1])\n",
    "        else:\n",
    "            hg38_coords.append(None)\n",
    "    dataset.iloc[:,1] = hg38_coords\n",
    "    dataset.to_csv(bedfile, header=None, index=None, sep='\\t')\n",
    "    \n",
    "def filter_accessible_vcfs(all_vcfs, output_bed,\n",
    "                           current_genome='hg19',\n",
    "                           chain_file='/home/amber/QuantPred/datasets/VCF/hg19ToHg38.over.chain',\n",
    "                          accessible_regions = 'combined_accessibility.bed'):\n",
    "    # if not provided make combined accessibility file\n",
    "    if not os.path.isfile(accessible_regions):\n",
    "        combine_accessible_regions(output_accessible_regions=accessible_regions)\n",
    "    if current_genome == 'hg19': # convert coordinates to hg38\n",
    "        convert_hg19_to_hg38(all_vcfs)\n",
    "    dataset = pd.read_csv(all_vcfs, header=None, sep='\\t')\n",
    "#     dataset\n",
    "    dataset.to_csv(all_vcfs, header=None, sep='\\t', index=None)\n",
    "    # intersect the hg38 coordinates with merged open chromatin region bed file for the current dataset\n",
    "    cmd = \"bedtools intersect -wa -a {} -b {} > vcf_open.bed;\".format(all_vcfs, \n",
    "                                                                    accessible_regions)\n",
    "    cmd += 'sort vcf_open.bed | uniq > {}'.format(output_bed)\n",
    "    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)\n",
    "    output, error = process.communicate()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_bed_to_seq(bedfile, output_fa, genomefile='/home/shush/genomes/hg38.fa'): \n",
    "    cmd = 'bedtools getfasta -fi {} -bed {} -s -fo {}'.format(genomefile, bedfile, output_fa)\n",
    "    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)\n",
    "    output, error = process.communicate()\n",
    "    coords_list, seqs_list = fasta2list(output_fa)\n",
    "    return coords_list, seqs_list\n",
    "\n",
    "def list_to_onehot(coords_list, seqs_list, nonneg_df):\n",
    "    N = len(seqs_list)\n",
    "    mid = window // 2\n",
    "    onehot_ref = np.empty((N, window, 4))\n",
    "    onehot_alt = np.empty((N, window, 4))\n",
    "    coord_np = np.empty((N, 4)) # chrom, start, end coordinate array\n",
    "    pos_dict = {'+': 1536, '-':1535}\n",
    "    for i,(chr_s_e, seq) in enumerate(zip(coords_list, seqs_list)):\n",
    "        alt = ''\n",
    "        strand = chr_s_e.split('(')[-1].split(')')[0]\n",
    "        pos = pos_dict[strand]\n",
    "        coord_np[i,3] = pos_dict[strand] - 1535\n",
    "\n",
    "        assert seq[pos] == nonneg_df.iloc[i, 3], 'reference allele error'\n",
    "        alt = nonneg_df.iloc[i,4]\n",
    "\n",
    "        chrom, s_e = chr_s_e.split('(')[0].split(':')\n",
    "        s, e = s_e.split('-')\n",
    "        coord_np[i, :3] = int(chrom.split('chr')[-1]), int(s), int(e)\n",
    "\n",
    "        onehot = dna_one_hot(seq)\n",
    "        onehot_ref[i,:,:] = onehot\n",
    "\n",
    "        onehot_alt[i, :, :] = onehot\n",
    "        onehot_alt[i, mid, :] = dna_one_hot(alt)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dsQTL dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = '.'\n",
    "window = 3072\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_paths = pd.read_csv('/mnt/906427d6-fddf-41bf-9ec6-c3d0c37e766f/amber/ATAC/basset_sample_file.tsv', sep='\\t', header=None)[1].values\n",
    "combined_csv = pd.concat([pd.read_csv(f, sep='\\t', header=None) for f in bed_paths ])\n",
    "combined_csv.to_csv('combined_atac.bed', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dsq_path = './dsQTL.csv'\n",
    "dsq_all = pd.read_csv(dsq_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snpChromStart</th>\n",
       "      <th>snpChromEnd</th>\n",
       "      <th>rsid</th>\n",
       "      <th>pred.fit.pctSig</th>\n",
       "      <th>strand</th>\n",
       "      <th>motifname</th>\n",
       "      <th>position</th>\n",
       "      <th>genotypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13971</td>\n",
       "      <td>13972</td>\n",
       "      <td>rs7536753,rs62530145,rs71259955</td>\n",
       "      <td>0.179</td>\n",
       "      <td>+</td>\n",
       "      <td>V_ETS_Q4</td>\n",
       "      <td>10</td>\n",
       "      <td>A/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13979</td>\n",
       "      <td>13980</td>\n",
       "      <td>rs2592732,rs62635296,rs151276478</td>\n",
       "      <td>0.109</td>\n",
       "      <td>-</td>\n",
       "      <td>V_AP1_C</td>\n",
       "      <td>1</td>\n",
       "      <td>G/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14906</td>\n",
       "      <td>14907</td>\n",
       "      <td>rs6682375,rs79585140</td>\n",
       "      <td>0.135</td>\n",
       "      <td>+</td>\n",
       "      <td>SPI1_ETS_1</td>\n",
       "      <td>11</td>\n",
       "      <td>A/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14947</td>\n",
       "      <td>14948</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0.110</td>\n",
       "      <td>-</td>\n",
       "      <td>V_AP2_Q6_01</td>\n",
       "      <td>3</td>\n",
       "      <td>T/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16256</td>\n",
       "      <td>16257</td>\n",
       "      <td>rs11489794,rs78588380</td>\n",
       "      <td>0.181</td>\n",
       "      <td>+</td>\n",
       "      <td>CTCF_UpstreamP1</td>\n",
       "      <td>30</td>\n",
       "      <td>C/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483410</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59027808</td>\n",
       "      <td>59027809</td>\n",
       "      <td>rs4047331</td>\n",
       "      <td>0.113</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD3_TEA_2</td>\n",
       "      <td>2</td>\n",
       "      <td>T/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483411</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59027861</td>\n",
       "      <td>59027862</td>\n",
       "      <td>rs76227494</td>\n",
       "      <td>0.104</td>\n",
       "      <td>+</td>\n",
       "      <td>V_TCF11MAFG_01</td>\n",
       "      <td>9</td>\n",
       "      <td>A/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483412</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59029164</td>\n",
       "      <td>59029165</td>\n",
       "      <td>rs79334518</td>\n",
       "      <td>0.151</td>\n",
       "      <td>-</td>\n",
       "      <td>V_CEBPB_02</td>\n",
       "      <td>10</td>\n",
       "      <td>G/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483413</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59029907</td>\n",
       "      <td>59029908</td>\n",
       "      <td>rs78084351</td>\n",
       "      <td>0.320</td>\n",
       "      <td>+</td>\n",
       "      <td>Jdp2.mouse_bZIP_1</td>\n",
       "      <td>5</td>\n",
       "      <td>A/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483414</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59029910</td>\n",
       "      <td>59029911</td>\n",
       "      <td>rs7892851,rs71229137</td>\n",
       "      <td>0.535</td>\n",
       "      <td>+</td>\n",
       "      <td>V_AP1_Q6</td>\n",
       "      <td>9</td>\n",
       "      <td>A/T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483415 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  snpChromStart  snpChromEnd                              rsid  \\\n",
       "0       chr1          13971        13972   rs7536753,rs62530145,rs71259955   \n",
       "1       chr1          13979        13980  rs2592732,rs62635296,rs151276478   \n",
       "2       chr1          14906        14907              rs6682375,rs79585140   \n",
       "3       chr1          14947        14948                       rs201855936   \n",
       "4       chr1          16256        16257             rs11489794,rs78588380   \n",
       "...      ...            ...          ...                               ...   \n",
       "483410  chrY       59027808     59027809                         rs4047331   \n",
       "483411  chrY       59027861     59027862                        rs76227494   \n",
       "483412  chrY       59029164     59029165                        rs79334518   \n",
       "483413  chrY       59029907     59029908                        rs78084351   \n",
       "483414  chrY       59029910     59029911              rs7892851,rs71229137   \n",
       "\n",
       "        pred.fit.pctSig strand          motifname  position genotypes  \n",
       "0                 0.179      +           V_ETS_Q4        10       A/C  \n",
       "1                 0.109      -            V_AP1_C         1       G/A  \n",
       "2                 0.135      +         SPI1_ETS_1        11       A/G  \n",
       "3                 0.110      -        V_AP2_Q6_01         3       T/C  \n",
       "4                 0.181      +    CTCF_UpstreamP1        30       C/G  \n",
       "...                 ...    ...                ...       ...       ...  \n",
       "483410            0.113      -        TEAD3_TEA_2         2       T/C  \n",
       "483411            0.104      +     V_TCF11MAFG_01         9       A/G  \n",
       "483412            0.151      -         V_CEBPB_02        10       G/A  \n",
       "483413            0.320      +  Jdp2.mouse_bZIP_1         5       A/C  \n",
       "483414            0.535      +           V_AP1_Q6         9       A/T  \n",
       "\n",
       "[483415 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dsq_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/amber/tf_2/lib/python3.7/site-packages/pandas/core/frame.py:3062: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    }
   ],
   "source": [
    "dsq_filt = dsq_all[(dsq_all['chrom']!='chrX')&(dsq_all['chrom']!='chrY' )]\n",
    "dsq_filt[['a1','a2']] = dsq_filt['genotypes'].str.split('/',expand=True) # write into separate columns\n",
    "dsq_filt.to_csv('dsq_all.bed', sep='\\t', header=False, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading liftover chains\r\n",
      "Mapping coordinates\r\n"
     ]
    }
   ],
   "source": [
    "!liftOver dsq_all.bed /home/amber/QuantPred/datasets/VCF/hg19ToHg38.over.chain dsq_hg38_all.bed unlifted.bed -bedPlus=3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools intersect -wa -a dsq_hg38_all.bed -b combined_atac.bed > dsq_open.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "! sort dsq_open.bed | uniq > dsq_unique.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import subprocess\n",
    "# def bed_intersect(dataset_bed='dsq_all.bed', comb_peak='combined_atac.bed', out_path='dsq_open.bed'):\n",
    "#     bashCmd = \"bedtools intersect -a {} -b {} > {}\".format(dataset_bed, comb_peak, out_path).split()\n",
    "#     process = subprocess.Popen(bashCmd, stdout=subprocess.PIPE)\n",
    "#     output, error = process.communicate()\n",
    "\n",
    "# bed_intersect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "dsq_df = pd.read_csv('dsq_unique.bed', sep='\\t', header=None, index_col=None)\n",
    "dsq_df.columns = list(dsq_filt)\n",
    "dsq_tmp = dsq_df[['chrom', 'snpChromStart', 'snpChromEnd', 'a1', 'a2', 'strand', 'rsid', 'pred.fit.pctSig']] # \n",
    "start, end = enforce_const_range(dsq_tmp['snpChromEnd']-1, window)\n",
    "dsq_ext = dsq_tmp.copy()\n",
    "dsq_ext.iloc[:,1] = start.values\n",
    "dsq_ext.iloc[:,2] = end.values\n",
    "dsq_nonneg = dsq_ext[dsq_ext['snpChromStart']>0]\n",
    "dsq_nonneg = dsq_nonneg.reset_index(drop=True)\n",
    "dsq_nonneg.to_csv('test_ds.csv', header=None, sep='\\t', index=None)\n",
    "!bedtools getfasta -fi /home/shush/genomes/hg38.fa -bed test_ds.csv -s -fo \"test_ds.fa\"\n",
    "\n",
    "coords_list, seqs_list = fasta2list('test_ds.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'bedtools getfasta -fi {} -bed {} -s -fo {}'.format(genomefile, bedfile, output_fa)\n",
    "process = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)\n",
    "output, error = process.communicate()\n",
    "coords_list, seqs_list = fasta2list(output_fa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'seqs_list' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-029278d803cb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mN\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseqs_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mmid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwindow\u001b[0m \u001b[0;34m//\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0monehot_ref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwindow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0monehot_alt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwindow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mcoord_np\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# chrom, start, end coordinate array\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'seqs_list' is not defined"
     ]
    }
   ],
   "source": [
    "N = len(seqs_list)\n",
    "mid = window // 2\n",
    "onehot_ref = np.empty((N, window, 4))\n",
    "onehot_alt = np.empty((N, window, 4))\n",
    "coord_np = np.empty((N, 4)) # chrom, start, end coordinate array\n",
    "\n",
    "\n",
    "\n",
    "for i,(chr_s_e, seq) in enumerate(zip(coords_list, seqs_list)):\n",
    "    alt = ''\n",
    "    strand = chr_s_e.split('(')[-1].split(')')[0]\n",
    "    pos_dict = {'+': 1536, '-':1535}\n",
    "    pos = pos_dict[strand]\n",
    "    coord_np[i,3] = pos_dict[strand] - 1535\n",
    "    \n",
    "    if seq[pos] == dsq_nonneg['a1'][i]:\n",
    "        alt = dsq_nonneg['a2'][i]\n",
    "        \n",
    "    elif seq[pos] == dsq_nonneg['a2'][i]:\n",
    "        alt = dsq_nonneg['a1'][i]\n",
    "    else:\n",
    "        break\n",
    "        print('reference allele error')\n",
    "\n",
    "    chrom, s_e = chr_s_e.split('(')[0].split(':')\n",
    "    s, e = s_e.split('-')\n",
    "    coord_np[i, :3] = int(chrom.split('chr')[-1]), int(s), int(e)\n",
    "    \n",
    "    onehot = dna_one_hot(seq)\n",
    "    onehot_ref[i,:,:] = onehot\n",
    "    \n",
    "    onehot_alt[i, :, :] = onehot\n",
    "    onehot_alt[i, mid, :] = dna_one_hot(alt)[0]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(135306, 3072, 4)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "onehot_ref.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array_equiv(onehot_ref, onehot_alt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "onehot_ref_alt = h5py.File(os.path.join(vcf_path, 'dsQTL_onehot.h5'), 'w')\n",
    "onehot_ref_alt.create_dataset('ref', data=onehot_ref)\n",
    "onehot_ref_alt.create_dataset('alt', data=onehot_alt)\n",
    "onehot_ref_alt.create_dataset('fasta_coords', data=coord_np)\n",
    "onehot_ref_alt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CAGI5 dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read df and add strand\n",
    "all_dfs = []\n",
    "cagi_data = 'CAGI'\n",
    "combined_filename = 'combined_cagi.bed'\n",
    "for filename in os.listdir(cagi_data):\n",
    "    prefix, regulator = filename.split('.tsv')[0].split('_')\n",
    "\n",
    "    one_reg = pd.read_csv(os.path.join(cagi_data,filename), skiprows=7, sep='\\t', header=None)\n",
    "    one_reg['regulator'] = regulator\n",
    "    one_reg['set'] = prefix\n",
    "    all_dfs.append(one_reg)\n",
    "    \n",
    "\n",
    "combined_cagi = pd.concat(all_dfs)\n",
    "combined_cagi.insert(4, 'strand', '+')\n",
    "combined_cagi.insert(2,'end',combined_cagi.iloc[:,1]+1)\n",
    "combined_cagi.iloc[:,0] = 'chr'+combined_cagi.iloc[:,0].astype(str)\n",
    "combined_cagi.to_csv(combined_filename, sep='\\t', header=False, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                chr10\n",
       "1             51548987\n",
       "end           51548988\n",
       "2                    C\n",
       "3                    A\n",
       "strand               +\n",
       "4                 0.02\n",
       "5                 0.01\n",
       "regulator         MSMB\n",
       "set          challenge\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_cagi.iloc[4339,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open_bedfile = 'cagi_open_vcf_unique.bed'\n",
    "# filter_accessible_vcfs(combined_filename, open_bedfile, current_genome='xxx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_filename = 'nonneg_cagi_3K.bed'\n",
    "expand_range(combined_filename, output_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa_filename = 'cagi_3k.fa'\n",
    "coords_list, seqs_list = convert_bed_to_seq(output_filename, fa_filename, genomefile='/home/shush/genomes/hg19.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 3072\n",
    "bad_lines = []\n",
    "N = len(seqs_list)\n",
    "nonneg_df = pd.read_csv(output_filename, sep='\\t', header=None)\n",
    "mid = window // 2\n",
    "onehot_ref = []\n",
    "onehot_alt = []\n",
    "coord_np = np.empty((N, 4)) # chrom, start, end coordinate array\n",
    "pos_dict = {'+': 1535, '-':1536}\n",
    "for i,(chr_s_e, seq) in enumerate(zip(coords_list, seqs_list)):\n",
    "    alt = ''\n",
    "    strand = chr_s_e.split('(')[-1].split(')')[0]\n",
    "    pos = pos_dict[strand]\n",
    "#     coord_np[i,3] = pos_dict[strand] - 1535\n",
    "\n",
    "    if seq[pos] != nonneg_df.iloc[i, 3]:\n",
    "#         print('Error in line ' + str(i))\n",
    "        bad_lines.append(i)\n",
    "    else:\n",
    "        alt = nonneg_df.iloc[i,4]\n",
    "\n",
    "        onehot = dna_one_hot(seq)\n",
    "        mutated_onehot = onehot.copy()\n",
    "        mutated_onehot[pos] = dna_one_hot(alt)[0]\n",
    "        onehot_ref.append(onehot)\n",
    "\n",
    "        onehot_alt.append(mutated_onehot) \n",
    "\n",
    "onehot_alt = np.array(onehot_alt)\n",
    "onehot_ref = np.array(onehot_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(onehot_alt == onehot_ref).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "included_df = nonneg_df[~nonneg_df.index.isin(bad_lines)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "included_df.to_csv('/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/shush/CAGI5/final_cagi_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5246732</td>\n",
       "      <td>5249804</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>HBB</td>\n",
       "      <td>release</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5246732</td>\n",
       "      <td>5249804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>HBB</td>\n",
       "      <td>release</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5246732</td>\n",
       "      <td>5249804</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>HBB</td>\n",
       "      <td>release</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5246733</td>\n",
       "      <td>5249805</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.17</td>\n",
       "      <td>HBB</td>\n",
       "      <td>release</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5246733</td>\n",
       "      <td>5249805</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>HBB</td>\n",
       "      <td>release</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18458</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11198687</td>\n",
       "      <td>11201759</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>0.04</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>challenge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18459</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11198687</td>\n",
       "      <td>11201759</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.04</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>challenge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18460</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11198688</td>\n",
       "      <td>11201760</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>challenge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18461</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11198688</td>\n",
       "      <td>11201760</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>challenge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18462</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11198688</td>\n",
       "      <td>11201760</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>challenge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18442 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2  3  4  5     6     7     8          9\n",
       "0      chr11   5246732   5249804  T  A  +  0.00  0.00   HBB    release\n",
       "1      chr11   5246732   5249804  T  C  +  0.02  0.01   HBB    release\n",
       "2      chr11   5246732   5249804  T  G  + -0.03  0.00   HBB    release\n",
       "3      chr11   5246733   5249805  G  A  + -0.16  0.17   HBB    release\n",
       "4      chr11   5246733   5249805  G  C  +  0.02  0.01   HBB    release\n",
       "...      ...       ...       ... .. .. ..   ...   ...   ...        ...\n",
       "18458  chr19  11198687  11201759  G  C  + -0.07  0.04  LDLR  challenge\n",
       "18459  chr19  11198687  11201759  G  T  +  0.07  0.04  LDLR  challenge\n",
       "18460  chr19  11198688  11201760  C  A  + -0.01  0.00  LDLR  challenge\n",
       "18461  chr19  11198688  11201760  C  G  +  0.02  0.01  LDLR  challenge\n",
       "18462  chr19  11198688  11201760  C  T  + -0.01  0.00  LDLR  challenge\n",
       "\n",
       "[18442 rows x 10 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "included_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "onehot_ref_alt = h5py.File('/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/shush/CAGI5/CAGI_onehot.h5', 'w')\n",
    "onehot_ref_alt.create_dataset('ref', data=onehot_ref)\n",
    "onehot_ref_alt.create_dataset('alt', data=onehot_alt)\n",
    "onehot_ref_alt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# negative eQTL dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 3072\n",
    "!grep negative ./eQTLs+negative.csv > negative.csv\n",
    "!cut -d\",\" -f2- negative.csv  > negative_clean.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "negative_table = pd.read_csv('negative_clean.csv', sep=',', skiprows=2,header =None,\n",
    "                            names = ['chrom','end','a1','a2','probability','fold','label'])\n",
    "negative_table['start'] = negative_table['end'] - 1\n",
    "negative_table = negative_table[['chrom','start','end','a1','a2','probability','fold','label']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "negative_table['hg19start'] = negative_table['start']\n",
    "negative_table['hg19end'] = negative_table['end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>probability</th>\n",
       "      <th>fold</th>\n",
       "      <th>label</th>\n",
       "      <th>hg19start</th>\n",
       "      <th>hg19end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10582</td>\n",
       "      <td>10583</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.35081</td>\n",
       "      <td>0</td>\n",
       "      <td>Random negative SNP</td>\n",
       "      <td>10582</td>\n",
       "      <td>10583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30922</td>\n",
       "      <td>30923</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14216</td>\n",
       "      <td>0</td>\n",
       "      <td>Random negative SNP</td>\n",
       "      <td>30922</td>\n",
       "      <td>30923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>52237</td>\n",
       "      <td>52238</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.18894</td>\n",
       "      <td>0</td>\n",
       "      <td>Random negative SNP</td>\n",
       "      <td>52237</td>\n",
       "      <td>52238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55163</td>\n",
       "      <td>55164</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.53187</td>\n",
       "      <td>0</td>\n",
       "      <td>Random negative SNP</td>\n",
       "      <td>55163</td>\n",
       "      <td>55164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88168</td>\n",
       "      <td>88169</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.48796</td>\n",
       "      <td>0</td>\n",
       "      <td>Random negative SNP</td>\n",
       "      <td>88168</td>\n",
       "      <td>88169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091033</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154891772</td>\n",
       "      <td>154891773</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.49647</td>\n",
       "      <td>9</td>\n",
       "      <td>360bp negative SNP</td>\n",
       "      <td>154891772</td>\n",
       "      <td>154891773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154900716</td>\n",
       "      <td>154900717</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.38994</td>\n",
       "      <td>9</td>\n",
       "      <td>360bp negative SNP</td>\n",
       "      <td>154900716</td>\n",
       "      <td>154900717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091035</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154913798</td>\n",
       "      <td>154913799</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.49817</td>\n",
       "      <td>9</td>\n",
       "      <td>360bp negative SNP</td>\n",
       "      <td>154913798</td>\n",
       "      <td>154913799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091036</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154925894</td>\n",
       "      <td>154925895</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.22989</td>\n",
       "      <td>9</td>\n",
       "      <td>360bp negative SNP</td>\n",
       "      <td>154925894</td>\n",
       "      <td>154925895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091037</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154930229</td>\n",
       "      <td>154930230</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.40443</td>\n",
       "      <td>9</td>\n",
       "      <td>360bp negative SNP</td>\n",
       "      <td>154930229</td>\n",
       "      <td>154930230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1091038 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end a1 a2  probability  fold  \\\n",
       "0        chr1      10582      10583  G  A      0.35081     0   \n",
       "1        chr1      30922      30923  G  T      0.14216     0   \n",
       "2        chr1      52237      52238  T  G      0.18894     0   \n",
       "3        chr1      55163      55164  C  A      0.53187     0   \n",
       "4        chr1      88168      88169  C  T      0.48796     0   \n",
       "...       ...        ...        ... .. ..          ...   ...   \n",
       "1091033  chrX  154891772  154891773  A  G      0.49647     9   \n",
       "1091034  chrX  154900716  154900717  C  A      0.38994     9   \n",
       "1091035  chrX  154913798  154913799  T  C      0.49817     9   \n",
       "1091036  chrX  154925894  154925895  C  T      0.22989     9   \n",
       "1091037  chrX  154930229  154930230  A  G      0.40443     9   \n",
       "\n",
       "                       label  hg19start    hg19end  \n",
       "0        Random negative SNP      10582      10583  \n",
       "1        Random negative SNP      30922      30923  \n",
       "2        Random negative SNP      52237      52238  \n",
       "3        Random negative SNP      55163      55164  \n",
       "4        Random negative SNP      88168      88169  \n",
       "...                      ...        ...        ...  \n",
       "1091033   360bp negative SNP  154891772  154891773  \n",
       "1091034   360bp negative SNP  154900716  154900717  \n",
       "1091035   360bp negative SNP  154913798  154913799  \n",
       "1091036   360bp negative SNP  154925894  154925895  \n",
       "1091037   360bp negative SNP  154930229  154930230  \n",
       "\n",
       "[1091038 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negative_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "negative_table.to_csv('./negative_all.bed', sep='\\t',header = None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading liftover chains\r\n",
      "Mapping coordinates\r\n"
     ]
    }
   ],
   "source": [
    "!liftOver negative_all.bed hg19ToHg38.over.chain negative_hg38_all.bed unlifted.bed -bedPlus=3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_negative_table = pd.read_csv('negative_hg38_all.bed', sep='\\t',header =None)\n",
    "all_negative_table = all_negative_table[(all_negative_table[7]=='Random')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10582</td>\n",
       "      <td>10583</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.35081</td>\n",
       "      <td>0</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>10582</td>\n",
       "      <td>10583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30922</td>\n",
       "      <td>30923</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14216</td>\n",
       "      <td>0</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>30922</td>\n",
       "      <td>30923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>52237</td>\n",
       "      <td>52238</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.18894</td>\n",
       "      <td>0</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>52237</td>\n",
       "      <td>52238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55163</td>\n",
       "      <td>55164</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.53187</td>\n",
       "      <td>0</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>55163</td>\n",
       "      <td>55164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88168</td>\n",
       "      <td>88169</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.48796</td>\n",
       "      <td>0</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>88168</td>\n",
       "      <td>88169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953976</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156000144</td>\n",
       "      <td>156000145</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.67814</td>\n",
       "      <td>9</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>155229809</td>\n",
       "      <td>155229810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953977</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156000160</td>\n",
       "      <td>156000161</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.62443</td>\n",
       "      <td>9</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>155229825</td>\n",
       "      <td>155229826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953978</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156003432</td>\n",
       "      <td>156003433</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.45205</td>\n",
       "      <td>9</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>155233097</td>\n",
       "      <td>155233098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953979</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156003449</td>\n",
       "      <td>156003450</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.36475</td>\n",
       "      <td>9</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>155233114</td>\n",
       "      <td>155233115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953980</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156006409</td>\n",
       "      <td>156006410</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.39778</td>\n",
       "      <td>9</td>\n",
       "      <td>Random</td>\n",
       "      <td>negative</td>\n",
       "      <td>SNP</td>\n",
       "      <td>155236074</td>\n",
       "      <td>155236075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>953981 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2  3  4        5   6       7         8    9   \\\n",
       "0       chr1      10582      10583  G  A  0.35081   0  Random  negative  SNP   \n",
       "1       chr1      30922      30923  G  T  0.14216   0  Random  negative  SNP   \n",
       "2       chr1      52237      52238  T  G  0.18894   0  Random  negative  SNP   \n",
       "3       chr1      55163      55164  C  A  0.53187   0  Random  negative  SNP   \n",
       "4       chr1      88168      88169  C  T  0.48796   0  Random  negative  SNP   \n",
       "...      ...        ...        ... .. ..      ...  ..     ...       ...  ...   \n",
       "953976  chrX  156000144  156000145  A  T  0.67814   9  Random  negative  SNP   \n",
       "953977  chrX  156000160  156000161  C  T  0.62443   9  Random  negative  SNP   \n",
       "953978  chrX  156003432  156003433  T  C  0.45205   9  Random  negative  SNP   \n",
       "953979  chrX  156003449  156003450  G  T  0.36475   9  Random  negative  SNP   \n",
       "953980  chrX  156006409  156006410  G  C  0.39778   9  Random  negative  SNP   \n",
       "\n",
       "               10         11  \n",
       "0           10582      10583  \n",
       "1           30922      30923  \n",
       "2           52237      52238  \n",
       "3           55163      55164  \n",
       "4           88168      88169  \n",
       "...           ...        ...  \n",
       "953976  155229809  155229810  \n",
       "953977  155229825  155229826  \n",
       "953978  155233097  155233098  \n",
       "953979  155233114  155233115  \n",
       "953980  155236074  155236075  \n",
       "\n",
       "[953981 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_negative_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "start, end = enforce_const_range(all_negative_table[2]-1,window)\n",
    "all_negative_extended_ranges = pd.DataFrame({'chr':all_negative_table[0],'start':start, 'end':end,\n",
    "                                    'a1':all_negative_table[3],'a2':all_negative_table[4],'hg19_start':all_negative_table[10] })\n",
    "sample_negative_range = all_negative_extended_ranges.sample(n=10000,replace=False)\n",
    "sample_negative_range.to_csv('negative_random_subsample_extended.csv', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>hg19_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>633323</th>\n",
       "      <td>chr4</td>\n",
       "      <td>97394607</td>\n",
       "      <td>97397679</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>98317294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88271</th>\n",
       "      <td>chr10</td>\n",
       "      <td>49118957</td>\n",
       "      <td>49122029</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>50328538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37987</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150305905</td>\n",
       "      <td>150308977</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>150279875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>874439</th>\n",
       "      <td>chr8</td>\n",
       "      <td>111631211</td>\n",
       "      <td>111634283</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>112644976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352892</th>\n",
       "      <td>chr17</td>\n",
       "      <td>67855384</td>\n",
       "      <td>67858456</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>65853036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164153</th>\n",
       "      <td>chr12</td>\n",
       "      <td>751938</td>\n",
       "      <td>755010</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>862640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918953</th>\n",
       "      <td>chr9</td>\n",
       "      <td>119197356</td>\n",
       "      <td>119200428</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>121961170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761234</th>\n",
       "      <td>chr6</td>\n",
       "      <td>105889947</td>\n",
       "      <td>105893019</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>106339358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741506</th>\n",
       "      <td>chr6</td>\n",
       "      <td>40388990</td>\n",
       "      <td>40392062</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>40358265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213909</th>\n",
       "      <td>chr13</td>\n",
       "      <td>30113888</td>\n",
       "      <td>30116960</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>30689561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr      start        end a1 a2  hg19_start\n",
       "633323   chr4   97394607   97397679  T  C    98317294\n",
       "88271   chr10   49118957   49122029  G  A    50328538\n",
       "37987    chr1  150305905  150308977  T  C   150279875\n",
       "874439   chr8  111631211  111634283  A  G   112644976\n",
       "352892  chr17   67855384   67858456  A  G    65853036\n",
       "...       ...        ...        ... .. ..         ...\n",
       "164153  chr12     751938     755010  T  G      862640\n",
       "918953   chr9  119197356  119200428  G  T   121961170\n",
       "761234   chr6  105889947  105893019  G  A   106339358\n",
       "741506   chr6   40388990   40392062  C  T    40358265\n",
       "213909  chr13   30113888   30116960  T  C    30689561\n",
       "\n",
       "[10000 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_negative_range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools getfasta -fi /home/shush/genomes/hg38.fa -bed './negative_random_subsample_extended.csv' -s -fo 'negative_random_subsample_extended.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "coords, nucl_seqs = fasta2list('negative_random_subsample_extended.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reference allele error362\n",
      "reference allele error538\n",
      "reference allele error1532\n",
      "reference allele error1731\n",
      "reference allele error2746\n",
      "reference allele error2842\n",
      "reference allele error3844\n",
      "reference allele error4079\n",
      "reference allele error4447\n",
      "reference allele error4714\n",
      "reference allele error6277\n",
      "reference allele error6423\n",
      "reference allele error6495\n",
      "reference allele error8444\n",
      "reference allele error8758\n",
      "reference allele error9931\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/amber/tf_2/lib/python3.7/site-packages/ipykernel_launcher.py:29: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    }
   ],
   "source": [
    "N = len(nucl_seqs)\n",
    "mid = window // 2\n",
    "onehot_ref = np.empty((N, window, 4))\n",
    "onehot_alt = np.empty((N, window, 4))\n",
    "\n",
    "for i in range(len(coords)):\n",
    "    \n",
    "    fa_allele = nucl_seqs[i][window//2]\n",
    "    if fa_allele == sample_negative_range['a1'].values[i]:\n",
    "        ref_allele = sample_negative_range['a1'].values[i]\n",
    "        alt_allele = sample_negative_range['a2'].values[i]\n",
    "    elif fa_allele == sample_negative_range['a2'].values[i]:\n",
    "        ref_allele = sample_negative_range['a2'].values[i]\n",
    "        alt_allele = sample_negative_range['a1'].values[i]\n",
    "    else:\n",
    "        print('reference allele error' + str(i))\n",
    "        next\n",
    "#     ref_allele = chr8_random_negative_table[3].values[i]\n",
    "#     alt_allele = chr8_random_negative_table[4].values[i]\n",
    "#     assert nucl_seqs[i][window//2] == ref_allele, str(i) + ' reference allele error'\n",
    "\n",
    "    onehot = dna_one_hot(nucl_seqs[i])\n",
    "    onehot_ref[i,:,:] = onehot\n",
    "    \n",
    "    onehot_alt[i, :, :] = onehot\n",
    "    onehot_alt[i, mid, :] = dna_one_hot(alt_allele)[0]\n",
    "\n",
    "\n",
    "coord_np = np.array([[8 for i in range(len(coords))], start.values, end.values]).T\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr6:61283848-61286920(61952329)'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coords[362]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr               chr6\n",
       "start         61283848\n",
       "end           61286920\n",
       "a1                   G\n",
       "a2                   T\n",
       "hg19_start    61952329\n",
       "Name: 747273, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_negative_range.iloc[362]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl_seqs[362][window//2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>hg19_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118697</th>\n",
       "      <td>chr11</td>\n",
       "      <td>3186192</td>\n",
       "      <td>3189264</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3208958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795919</th>\n",
       "      <td>chr7</td>\n",
       "      <td>29543878</td>\n",
       "      <td>29546950</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>29585030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169707</th>\n",
       "      <td>chr12</td>\n",
       "      <td>12971032</td>\n",
       "      <td>12974104</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>13125502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286729</th>\n",
       "      <td>chr15</td>\n",
       "      <td>62983620</td>\n",
       "      <td>62986692</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>63277355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567090</th>\n",
       "      <td>chr3</td>\n",
       "      <td>100898392</td>\n",
       "      <td>100901464</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>100618772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437752</th>\n",
       "      <td>chr2</td>\n",
       "      <td>73254392</td>\n",
       "      <td>73257464</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>73483056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313391</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29116523</td>\n",
       "      <td>29119595</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>29129380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265083</th>\n",
       "      <td>chr14</td>\n",
       "      <td>89493650</td>\n",
       "      <td>89496722</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>89961530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777585</th>\n",
       "      <td>chr6</td>\n",
       "      <td>158224817</td>\n",
       "      <td>158227889</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>158647385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297760</th>\n",
       "      <td>chr15</td>\n",
       "      <td>94649358</td>\n",
       "      <td>94652430</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>95194123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr      start        end a1 a2  hg19_start\n",
       "118697  chr11    3186192    3189264  A  G     3208958\n",
       "795919   chr7   29543878   29546950  C  T    29585030\n",
       "169707  chr12   12971032   12974104  G  C    13125502\n",
       "286729  chr15   62983620   62986692  T  C    63277355\n",
       "567090   chr3  100898392  100901464  T  C   100618772\n",
       "...       ...        ...        ... .. ..         ...\n",
       "437752   chr2   73254392   73257464  T  C    73483056\n",
       "313391  chr16   29116523   29119595  A  C    29129380\n",
       "265083  chr14   89493650   89496722  G  A    89961530\n",
       "777585   chr6  158224817  158227889  C  A   158647385\n",
       "297760  chr15   94649358   94652430  A  C    95194123\n",
       "\n",
       "[10000 rows x 6 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_negative_range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = './'\n",
    "onehot_ref_alt = h5py.File(os.path.join(vcf_path, 'negative_onehot.h5'), 'w')\n",
    "onehot_ref_alt.create_dataset('ref', data=onehot_ref)\n",
    "onehot_ref_alt.create_dataset('alt', data=onehot_alt)\n",
    "onehot_ref_alt.create_dataset('fasta_coords', data=coord_np)\n",
    "onehot_ref_alt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try background seq generation in hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 3072\n",
    "!grep negative ./eQTLs+negative.csv > negative.csv\n",
    "!cut -d\",\" -f2- negative.csv  > negative_clean.csv\n",
    "negative_table = pd.read_csv('negative_clean.csv', sep=',', skiprows=2,header =None,\n",
    "                            names = ['chrom','end','a1','a2','probability','fold','label'])\n",
    "negative_table['start'] = negative_table['end'] - 1\n",
    "negative_table = negative_table[['chrom','start','end','a1','a2','probability','fold','label']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_negative_table = negative_table[(negative_table['label'] == 'Random negative SNP')&\n",
    "                                    (negative_table['chrom']!='chrX')&(negative_table['chrom']!='chrY')]\n",
    "start, end = enforce_const_range(all_negative_table['end']-1,window)\n",
    "all_negative_extended_ranges = pd.DataFrame({'chr':all_negative_table['chrom'],'start':start, 'end':end,\n",
    "                                    'a1':all_negative_table['a1'],'a2':all_negative_table['a2']})\n",
    "sample_negative_range = all_negative_extended_ranges.sample(n=10000,replace=False)\n",
    "sample_negative_range.to_csv('negative_random_subsample_extended.csv', header=None, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools getfasta -fi /home/shush/genomes/hg19.fa -bed './negative_random_subsample_extended.csv' -s -fo 'negative_random_subsample_extended.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "coords, nucl_seqs = fasta2list('negative_random_subsample_extended.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr9:24623578-24626650()'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coords[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = len(nucl_seqs)\n",
    "mid = window // 2\n",
    "onehot_ref = np.empty((N, window, 4))\n",
    "onehot_alt = np.empty((N, window, 4))\n",
    "coord_np = np.empty((N, 4))\n",
    "\n",
    "for i in range(len(coords)):\n",
    "    \n",
    "    fa_allele = nucl_seqs[i][window//2]\n",
    "    if fa_allele == sample_negative_range['a1'].values[i]:\n",
    "        ref_allele = sample_negative_range['a1'].values[i]\n",
    "        alt_allele = sample_negative_range['a2'].values[i]\n",
    "    elif fa_allele == sample_negative_range['a2'].values[i]:\n",
    "        ref_allele = sample_negative_range['a2'].values[i]\n",
    "        alt_allele = sample_negative_range['a1'].values[i]\n",
    "    else:\n",
    "        print('reference allele error' + str(i))\n",
    "        next\n",
    "#     ref_allele = chr8_random_negative_table[3].values[i]\n",
    "#     alt_allele = chr8_random_negative_table[4].values[i]\n",
    "#     assert nucl_seqs[i][window//2] == ref_allele, str(i) + ' reference allele error'\n",
    "\n",
    "    onehot = dna_one_hot(nucl_seqs[i])\n",
    "    onehot_ref[i,:,:] = onehot\n",
    "    \n",
    "    onehot_alt[i, :, :] = onehot\n",
    "    onehot_alt[i, mid, :] = dna_one_hot(alt_allele)[0]\n",
    "\n",
    "    \n",
    "    chr_s_e = coords[i]\n",
    "    chrom, s_e = chr_s_e.split('(')[0].split(':')\n",
    "    s, e = s_e.split('-')\n",
    "    coord_np[i, :3] = int(chrom.split('chr')[-1]), int(s), int(e)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = './'\n",
    "onehot_ref_alt = h5py.File(os.path.join(vcf_path, 'negative_onehot.h5'), 'w')\n",
    "onehot_ref_alt.create_dataset('ref', data=onehot_ref)\n",
    "onehot_ref_alt.create_dataset('alt', data=onehot_alt)\n",
    "onehot_ref_alt.create_dataset('fasta_coords', data=coord_np)\n",
    "onehot_ref_alt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "onehot_ref_alt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0b3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
