{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook goes through dataset creation for quantitative models. \n",
    "**!!!** If not running on bizon or not using GRCh38 or mm9 genomes please download fasta files and change paths in the set_genome function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**required scripts for running this notebook:**\n",
    "- /home/shush/profile/QuantPred/datasets/basenji_data.py\n",
    "- /home/shush/profile/QuantPred/datasets/basenji_data_read.py\n",
    "- /home/shush/profile/QuantPred/datasets/basenji_data_write.py\n",
    "- /home/shush/profile/QuantPred/datasets/genome.py\n",
    "- /home/shush/profile/QuantPred/datasets/util.py\n",
    "- /home/shush/profile/QuantPred/datasets/dna_io.py\n",
    "\n",
    "or just pull git@github.com:amberT15/QuantPred.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import subprocess\n",
    "import yaml, os, shutil, sys\n",
    "import util\n",
    "import json\n",
    "from natsort import natsorted\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_genome(genome): \n",
    "    '''choose genome file paths for chr size, fa and unmappable genome segments (optional)'''\n",
    "    genome_dict = {'hg38': {'size':  '/home/shush/genomes/GRCh38_EBV.chrom.sizes.tsv',\n",
    "                              'fa':  '/home/shush/genomes/hg38.fa',\n",
    "                              'unmap':  '/home/shush/genomes/GRCh38_unmap.bed'},\n",
    "                  'mm9': {'size':  '/home/shush/genomes/mm9.chrom.sizes',\n",
    "                          'fa':  '/home/shush/genomes/mm9.fa',\n",
    "                          'unmap':  '/home/shush/genomes/mm9-blacklist.bed'}}\n",
    "            \n",
    "    assert genome in genome_dict.keys(), 'Unknown genome!'\n",
    "    return genome_dict[genome]\n",
    "\n",
    "def write_basenji_samplefile(bigwig_filepaths, basenji_samplefile):\n",
    "    df = pd.DataFrame(columns =['index', 'identifier', 'file', 'sum_stat', 'description'])\n",
    "    # per file, get the filename\n",
    "    for b, bigwig_filepath in enumerate(bigwig_filepaths):\n",
    "        # make entry in basenji samplefile\n",
    "        df.loc[b] = [b, os.path.basename(bigwig_filepath).split('.b')[0], bigwig_filepath, 'mean', '']\n",
    "\n",
    "    # write to csv files\n",
    "    df.to_csv(basenji_samplefile, index=None, sep='\\t')\n",
    "    \n",
    "def write_basset_samplefile(bed_filepaths, basset_samplefile):\n",
    "    print('Generating merged samplefile for the entire bedfile set')\n",
    "    df = pd.DataFrame(columns =['identifier', 'file'])\n",
    "    # per file, get the filename\n",
    "    for b, bedfile_path in enumerate(bed_filepaths):\n",
    "        # make entry in basenji samplefile\n",
    "        df.loc[b] = [os.path.basename(bedfile_path).split('.b')[0], bedfile_path]\n",
    "    # write to csv files\n",
    "    df.to_csv(basset_samplefile, index=None, header=None, sep='\\t')\n",
    "    \n",
    "def make_dataset(data_dir, split_label, data_stats, batch_size=64, seed=None, shuffle=True, coords=False):\n",
    "    seq_length = data_stats['seq_length']\n",
    "    target_length = data_stats['target_length']\n",
    "    num_targets = data_stats['num_targets']\n",
    "    tfr_path = '%s/tfrecords/%s-*.tfr' % (data_dir, split_label)\n",
    "    num_seqs = data_stats['%s_seqs' % split_label]\n",
    "\n",
    "    tfr_files = natsorted(glob.glob(tfr_path))\n",
    "    dataset = tf.data.Dataset.list_files(tf.constant(tfr_files), shuffle=False)\n",
    "\n",
    "    # train\n",
    "    # if split_label == 'train':\n",
    "    if (split_label == 'train'):\n",
    "      # repeat\n",
    "      #dataset = dataset.repeat()\n",
    "\n",
    "      # interleave files\n",
    "      dataset = dataset.interleave(map_func=file_to_records,\n",
    "        cycle_length=4,\n",
    "        num_parallel_calls=tf.data.experimental.AUTOTUNE)\n",
    "\n",
    "      # shuffle\n",
    "      dataset = dataset.shuffle(buffer_size=32,\n",
    "        reshuffle_each_iteration=True)\n",
    "\n",
    "    # valid/test\n",
    "    else:\n",
    "      # flat mix files\n",
    "      dataset = dataset.flat_map(file_to_records)\n",
    "\n",
    "    dataset = dataset.map(generate_parser(seq_length, target_length, num_targets, coords))\n",
    "    if shuffle:\n",
    "        if seed:\n",
    "            dataset = dataset.shuffle(32, seed=seed)\n",
    "        else:\n",
    "            dataset = dataset.shuffle(32)\n",
    "    # dataset = dataset.batch(64)\n",
    "    # batch\n",
    "    dataset = dataset.batch(batch_size)\n",
    "\n",
    "    # prefetch\n",
    "    dataset = dataset.prefetch(tf.data.experimental.AUTOTUNE)\n",
    "    return dataset\n",
    "\n",
    "def load_stats(data_dir):\n",
    "    data_stats_file = '%s/statistics.json' % data_dir\n",
    "    with open(data_stats_file) as data_stats_open:\n",
    "        data_stats = json.load(data_stats_open)\n",
    "    return data_stats\n",
    "\n",
    "def file_to_records(filename):\n",
    "    return tf.data.TFRecordDataset(filename, compression_type='ZLIB')\n",
    "\n",
    "def generate_parser(seq_length, target_length, num_targets, coords):\n",
    "    def parse_proto(example_protos):\n",
    "        \"\"\"Parse TFRecord protobuf.\"\"\"\n",
    "        # TFRecord constants\n",
    "        TFR_COORD = 'coordinate'\n",
    "        TFR_INPUT = 'sequence'\n",
    "        TFR_OUTPUT = 'target'\n",
    "\n",
    "        # define features\n",
    "        features = {\n",
    "          TFR_COORD: tf.io.FixedLenFeature([], tf.string),\n",
    "          TFR_INPUT: tf.io.FixedLenFeature([], tf.string),\n",
    "          TFR_OUTPUT: tf.io.FixedLenFeature([], tf.string)\n",
    "        }\n",
    "\n",
    "        # parse example into features\n",
    "        parsed_features = tf.io.parse_single_example(example_protos, features=features)\n",
    "\n",
    "        # decode coords\n",
    "        coordinate = parsed_features[TFR_COORD]\n",
    "\n",
    "        # decode sequence\n",
    "        # sequence = tf.io.decode_raw(parsed_features[TFR_INPUT], tf.uint8)\n",
    "        sequence = tf.io.decode_raw(parsed_features[TFR_INPUT], tf.float16)\n",
    "        sequence = tf.reshape(sequence, [seq_length, 4])\n",
    "        sequence = tf.cast(sequence, tf.float32)\n",
    "\n",
    "        # decode targets\n",
    "        targets = tf.io.decode_raw(parsed_features[TFR_OUTPUT], tf.float16)\n",
    "        targets = tf.reshape(targets, [target_length, num_targets])\n",
    "        targets = tf.cast(targets, tf.float32)\n",
    "        if coords:\n",
    "            return coordinate, sequence, targets\n",
    "        else:\n",
    "            return sequence, targets\n",
    "\n",
    "    return parse_proto\n",
    "\n",
    "def convert_tfr_to_np(testset, number_data_types=2):\n",
    "    all_data = [[] for i in range(number_data_types)]\n",
    "    for i, (data) in enumerate(testset):\n",
    "        for j, data_type in enumerate(data):\n",
    "            all_data[j].append(data_type)\n",
    "    return [np.concatenate(d) for d in all_data]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The most relevant options here are:\n",
    "\n",
    "| Argument | Note |\n",
    "|:---|:---|\n",
    "| chroms_only | if 'all' creates train, val, test, if specific chromosomes then creates test set from only those|\n",
    "| input_size| input size of the genomic|\n",
    "| base_dir | the output directory for the tfr files |\n",
    "| bigwig_paths_pattern | regexp pattern that will collect all the bigwig files |\n",
    "| bigwig_filepaths | can set this instead as a list of all the bws if don't want to use glob |\n",
    "| bedfile_paths_pattern | same for bed files of genomic regions to focus on IF you want peak centered dataset |\n",
    "| bed_filepaths | --.-- |\n",
    "| pool_window | bin size, if set to 1 can bin later in the training |\n",
    "| dilation_rate | fraction of data to include, can set to 0.1 to test pipelines|\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making directory: /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc\n"
     ]
    }
   ],
   "source": [
    "input_size = 3072\n",
    "base_dir = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc'\n",
    "bigwig_paths_pattern = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/*/Bigwig/*'\n",
    "bigwig_filepaths = [f for f in glob.glob(bigwig_paths_pattern) if f.endswith('bigwig') or f.endswith('bw')]\n",
    "bedfile_paths_pattern = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/*/Bed/*'\n",
    "bed_filepaths = [f for f in glob.glob(bedfile_paths_pattern) if f.endswith('bed') or f.endswith('gz')]\n",
    "\n",
    "# optional arguments\n",
    "pool_window = 1 \n",
    "dilation_rate = 1 \n",
    "valid_chr = 'chr9'\n",
    "test_chr = 'chr8'\n",
    "genome = 'mm9'\n",
    "chroms_only = 'all'\n",
    "norm = 'none'\n",
    "threshold = 0 # threshold for train and val\n",
    "test_threshold = 0 # test threshold\n",
    "# only change these if you want overlapping inputs\n",
    "step = 0\n",
    "stride_test = 1 #e.g. set to 0.333333333333 to move by 1K if 3K input\n",
    "padding = 'none'\n",
    "util.make_directory(base_dir) # create dir if not there already\n",
    "# basset_samplefile = os.path.join(base_dir, 'basset_samplefile.csv') # change to random for thresholded\n",
    "basset_samplefile = 'random'\n",
    "basenji_samplefile = os.path.join(base_dir, 'basenji_samplefile.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set([os.path.basename(b) for b in bigwig_filepaths]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating merged samplefile for the entire bedfile set\n"
     ]
    }
   ],
   "source": [
    "\n",
    "write_basenji_samplefile(bigwig_filepaths, basenji_samplefile) # write pre-requisite file for the pipeline specifying bw paths\n",
    "\n",
    "if basset_samplefile != 'random':\n",
    "    write_basset_samplefile(bed_filepaths, basset_samplefile)# write pre-requisite file for the pipeline specifying bed paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_path = os.path.join(base_dir, 'config.yaml')\n",
    "config = {}\n",
    "\n",
    "config['genomefile'] = set_genome(genome)\n",
    "\n",
    "config['chroms']={'valid': valid_chr, 'test': test_chr, 'only': chroms_only}\n",
    "config['input'] = {'downsample': dilation_rate, 'size':input_size, 'pool':pool_window, \n",
    "                   'norm': norm, 'step':step, 'padding':padding}\n",
    "config['samplefile'] = {'basset': basset_samplefile, 'basenji':basenji_samplefile}\n",
    "\n",
    "config['threshold'] = threshold\n",
    "config['test_threshold'] = test_threshold\n",
    "config['output'] = {'dir': base_dir, \n",
    "                   'prefix': 'i_%i_w_%i' % (config['input']['size'], config['input']['pool'])}\n",
    "config['stride_test'] = stride_test "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('config.yaml', 'w') as file:\n",
    "    documents = yaml.dump(config, file, default_flow_style=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!!!\n",
      "/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc\n",
      "/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/basset_samplefile.csv\n",
      "Peak centering\n",
      "Generating bed region combined file for all TFs\n",
      "Sorting bedfile and genome file\n",
      "Generating bed file complementary to peak regions\n",
      "Merging nonpeak and blacklisted regions\n",
      "Running basenji data processing\n",
      "/home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/basenji_samplefile.csv -g merged_avoid_regions.bed -l 3072 -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 -t chr8 -v chr9 -w 1 --local -d 0.1 --norm none --step 0 --padding none -p 19 --threshold 0 --test_threshold 0 --only_chroms all --stride_test 1\n",
      "['/home/shush/genomes/mm9.fa', '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/basenji_samplefile.csv']\n",
      "Using test set threshold of 0.0\n",
      "stride_train 1 converted to 3072.000000\n",
      "stride_test 1 converted to 3072.000000\n",
      "CHR\n",
      "Contigs divided into\n",
      " Train: 145518 contigs,  454666569 nt (0.8954)\n",
      " Valid:  8566 contigs,   26760458 nt (0.0527)\n",
      " Test:   8430 contigs,   26360231 nt (0.0519)\n",
      "<generator object divide_contigs_chr.<locals>.<genexpr> at 0x7f7d7e4bb228>\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/0.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/1.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Rep1_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/2.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Rep1_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/3.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Rep2_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/4.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/5.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/6.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Rep2_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/7.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD5_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/8.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD2_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/9.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD2_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/10.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD5_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/11.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/12.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_Parous_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/13.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_During_2ndPreg_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/14.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_During_1stPreg_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov/15.h5\n",
      "./basenji_data_write.py -s 0 -e 256 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-0.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 256 -e 512 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-1.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 512 -e 768 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-2.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 768 -e 1024 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-3.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 1024 -e 1280 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-4.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 1280 -e 1536 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-5.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./basenji_data_write.py -s 1536 -e 1792 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-6.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 1792 -e 2048 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-7.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 2048 -e 2304 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-8.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 2304 -e 2560 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-9.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 2560 -e 2816 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-10.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 2816 -e 3072 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-11.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 3072 -e 3328 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-12.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 3328 -e 3584 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-13.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 3584 -e 3840 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-14.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 3840 -e 4096 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-15.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 4096 -e 4352 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-16.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 4352 -e 4608 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-17.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 4608 -e 4864 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-18.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "./basenji_data_write.py -s 4864 -e 5120 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-19.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 5120 -e 5376 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-20.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 5376 -e 5632 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-21.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 5632 -e 5888 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-22.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 5888 -e 6144 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-23.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 6144 -e 6400 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-24.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 6400 -e 6656 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-25.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 6656 -e 6912 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-26.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 6912 -e 7168 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-27.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 7168 -e 7424 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-28.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 7424 -e 7680 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-29.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 7680 -e 7936 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-30.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 7936 -e 8192 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-31.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 8192 -e 8448 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-32.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 8448 -e 8704 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-33.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./basenji_data_write.py -s 8704 -e 8960 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-34.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 8960 -e 9216 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-35.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 9216 -e 9472 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-36.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 9472 -e 9728 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-37.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "./basenji_data_write.py -s 9728 -e 9984 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-38.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 9984 -e 10240 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-39.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 10240 -e 10496 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-40.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 10496 -e 10752 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-41.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 10752 -e 11008 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-42.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 11008 -e 11264 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-43.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 11264 -e 11520 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-44.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 11520 -e 11776 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-45.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 11776 -e 12032 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-46.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 12032 -e 12288 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-47.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 12288 -e 12544 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-48.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./basenji_data_write.py -s 12544 -e 12800 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-49.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 12800 -e 13056 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-50.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 13056 -e 13312 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-51.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 13312 -e 13568 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-52.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 13568 -e 13824 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-53.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 13824 -e 14080 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-54.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 14080 -e 14336 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-55.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 14336 -e 14557 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/train-56.tfr train -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "221\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "./basenji_data_write.py -s 14557 -e 14813 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/valid-0.tfr valid -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 14813 -e 15069 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/valid-1.tfr valid -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 15069 -e 15325 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/valid-2.tfr valid -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 15325 -e 15414 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/valid-3.tfr valid -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "./basenji_data_write.py -s 15414 -e 16257 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/tfrecords/test-0.tfr test -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1 --threshold 0.000000 --test_threshold 0.000000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "89\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "256\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "843\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset_pc/i_3072_w_1/statistics.json\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n"
     ]
    }
   ],
   "source": [
    "! ./bw_to_tfr.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BalbC_Virgin_27ac_q1e-03\t/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bed/BalbC_Virgin_27ac_q1e-03.bed\r\n",
      "BalbC_During_2ndPreg_27ac_q1e-03\t/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bed/BalbC_During_2ndPreg_27ac_q1e-03.bed\r\n",
      "BalbC_During_1stPreg_27ac_q1e-03\t/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bed/BalbC_During_1stPreg_27ac_q1e-03.bed\r\n",
      "BalbC_Parous_27a_q1e-03\t/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bed/BalbC_Parous_27a_q1e-03.bed\r\n"
     ]
    }
   ],
   "source": [
    "! head /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset/basenji_samplefile.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  42787 test\n",
      " 717678 train\n",
      "  40239 valid\n"
     ]
    }
   ],
   "source": [
    "# sanity check for dataset splits\n",
    "! cut -f4 /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset/i_3072_w_1/sequences.bed | sort | uniq -c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chroms:\r\n",
      "  only: all\r\n",
      "  test: chr8\r\n",
      "  valid: chr9\r\n",
      "genomefile:\r\n",
      "  fa: /home/shush/genomes/mm9.fa\r\n",
      "  size: /home/shush/genomes/mm9.chrom.sizes\r\n",
      "  unmap: /home/shush/genomes/mm9-blacklist.bed\r\n",
      "input:\r\n",
      "  downsample: 1\r\n",
      "  norm: none\r\n",
      "  padding: none\r\n",
      "  pool: 1\r\n",
      "  size: 3072\r\n",
      "  step: 0\r\n",
      "output:\r\n",
      "  dir: /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset\r\n",
      "  prefix: i_3072_w_1\r\n",
      "samplefile:\r\n",
      "  basenji: /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset/basenji_samplefile.csv\r\n",
      "  basset: random\r\n",
      "stride_test: 1\r\n",
      "test_threshold: -1\r\n",
      "threshold: 2\r\n"
     ]
    }
   ],
   "source": [
    "# summary of the configs corresponding to the new dataset\n",
    "! cat /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset/i_3072_w_1/config.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data is now in tfrecords which is optimized for faster i/o and speeds up training. But if needed it can easily be converted to numpy arrays. Here I look at the test set only and retireve the sequences, targets and the coordinates as well (which is why we need `coords=True, number_data_types=3`). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/tfr_dataset/i_3072_w_1/'\n",
    "sts = load_stats(data_dir)\n",
    "testset = make_dataset(data_dir, 'test', sts, batch_size=32, shuffle=False, coords=True)\n",
    "targets = pd.read_csv(data_dir+'targets.txt', sep='\\t')['identifier'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CAGMYC_DD5_Parous_Input_27ac', 'CAGMYC_DD5_Virgin_Input_27ac',\n",
       "       'CAGMYC_DD5_Parous_Rep1_ATAC', 'CAGMYC_DD5_Virgin_Rep1_ATAC',\n",
       "       'CAGMYC_DD5_Parous_Rep2_ATAC', 'CAGMYC_DD5_Virgin_27ac',\n",
       "       'CAGMYC_DD5_Parous_27ac', 'CAGMYC_DD5_Virgin_Rep2_ATAC',\n",
       "       'CAGMYC_DD5_Virgin_Input_27ac', 'CAGMYC_DD2_Virgin_27ac',\n",
       "       'CAGMYC_DD2_Virgin_Input_27ac', 'CAGMYC_DD5_Virgin_27ac',\n",
       "       'BalbC_Virgin_27ac', 'BalbC_Parous_27ac',\n",
       "       'BalbC_During_2ndPreg_27ac', 'BalbC_During_1stPreg_27ac'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<DatasetV1Adapter shapes: ((None,), (None, 3072, 4), (None, 3072, 16)), types: (tf.string, tf.float32, tf.float32)>"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "C, X, Y = convert_tfr_to_np(testset, number_data_types=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b'chr8_3072000_3075072'\n"
     ]
    }
   ],
   "source": [
    "print(C[1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make just the test set with a 2K window size for evaluation\n",
    "\n",
    "this is useful if you already have a dataset but want to make a smaller one only from specific chromosomes and potentially with other thresholds, peak files, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making directory: /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8\n"
     ]
    }
   ],
   "source": [
    "input_size = 2048\n",
    "base_dir = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8'\n",
    "bigwig_paths_pattern = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/*/Bigwig/*'\n",
    "bigwig_filepaths = [f for f in glob(bigwig_paths_pattern) if f.endswith('bigwig') or f.endswith('bw')]\n",
    "bedfile_paths_pattern = '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/*/Bed/*'\n",
    "bed_filepaths = [f for f in glob(bedfile_paths_pattern) if f.endswith('bed') or f.endswith('gz')]\n",
    "\n",
    "# optional arguments\n",
    "pool_window = 1 \n",
    "dilation_rate =0.1 \n",
    "valid_chr = 'chr9'\n",
    "test_chr = 'chr8'\n",
    "genome = 'mm9'\n",
    "chroms_only = 'chr8' #*********** <------ ********* change this to comma separated list of chr you want to include\n",
    "norm = 'none'\n",
    "threshold = 0 # threshold for train and val\n",
    "test_threshold = 0 # test threshold\n",
    "# only change these if you want overlapping inputs\n",
    "step = 0\n",
    "stride_test = 1 #e.g. set to 0.333333333333 to move by 1K if 3K input\n",
    "padding = 'none'\n",
    "util.make_directory(base_dir) # create dir if not there already\n",
    "basset_samplefile =  'random' #os.path.join(base_dir, 'basset_samplefile.csv') \n",
    "basenji_samplefile = os.path.join(base_dir, 'basenji_samplefile.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "write_basenji_samplefile(bigwig_filepaths, basenji_samplefile) # write pre-requisite file for the pipeline specifying bw paths\n",
    "\n",
    "if basset_samplefile != 'random':\n",
    "    write_basset_samplefile(bed_filepaths, basset_samplefile)# write pre-requisite file for the pipeline specifying bed paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "config_path = os.path.join(base_dir, 'config.yaml')\n",
    "config = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_path = os.path.join(base_dir, 'config.yaml')\n",
    "config = {}\n",
    "\n",
    "config['genomefile'] = set_genome(genome)\n",
    "\n",
    "config['chroms']={'valid': valid_chr, 'test': test_chr, 'only': chroms_only}\n",
    "config['input'] = {'downsample': dilation_rate, 'size':input_size, 'pool':pool_window, \n",
    "                   'norm': norm, 'step':step, 'padding':padding}\n",
    "config['samplefile'] = {'basset': basset_samplefile, 'basenji':basenji_samplefile}\n",
    "\n",
    "config['threshold'] = threshold\n",
    "config['test_threshold'] = test_threshold\n",
    "config['output'] = {'dir': base_dir, \n",
    "                   'prefix': 'i_%i_w_%i' % (config['input']['size'], config['input']['pool'])}\n",
    "config['stride_test'] = stride_test "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('config.yaml', 'w') as file:\n",
    "    documents = yaml.dump(config, file, default_flow_style=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!!!\n",
      "/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8\n",
      "random\n",
      "Chopping randomly\n",
      "Running basenji data processing\n",
      "/home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/basenji_samplefile.csv -g /home/shush/genomes/mm9-blacklist.bed -l 2048 -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1 -t chr8 -v chr9 -w 1 --local -d 0.1 --norm none --step 0 --padding none -p 19 --threshold 0 --test_threshold 0 --only_chroms chr8 --stride_test 1\n",
      "['/home/shush/genomes/mm9.fa', '/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/basenji_samplefile.csv']\n",
      "Using test set threshold of 0.0\n",
      "stride_train 1 converted to 2048.000000\n",
      "stride_test 1 converted to 2048.000000\n",
      "chr8\n",
      "CHR\n",
      "Contigs divided into\n",
      " Train:     0 contigs,          0 nt (0.0000)\n",
      " Valid:     0 contigs,          0 nt (0.0000)\n",
      " Test:     18 contigs,  131470271 nt (1.0000)\n",
      "<generator object divide_contigs_chr.<locals>.<genexpr> at 0x7fdadafe9228>\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/0.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/1.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Rep1_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/2.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Rep1_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/3.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_Rep2_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/4.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/5.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Parous_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/6.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/C/Bigwig/CAGMYC_DD5_Virgin_Rep2_ATAC.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/7.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD5_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/8.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD2_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/9.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD2_Virgin_Input_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/10.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/B/Bigwig/CAGMYC_DD5_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/11.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_Virgin_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/12.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_Parous_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/13.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_During_2ndPreg_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/14.h5\n",
      "./basenji_data_read.py --crop 0 -w 1 -u mean -s 1.000000 --norm none --padding none /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/datasets/A/Bigwig/BalbC_During_1stPreg_27ac.bigwig /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov/15.h5\n",
      "./basenji_data_write.py -s 0 -e 6418 --umap_clip 1.000000 /home/shush/genomes/mm9.fa /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/seqs_cov /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/tfrecords/test-0.tfr test -o /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1 --threshold 0.000000 --test_threshold 0.000000\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "4954\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "./basenji_data_write.py:232: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  values = values.flatten().tostring()\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "/mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/statistics.json\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n",
      "~~~\n"
     ]
    }
   ],
   "source": [
    "! ./bw_to_tfr.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   6418 test\r\n"
     ]
    }
   ],
   "source": [
    "! cut -f4 /mnt/31dac31c-c4e2-4704-97bd-0788af37c5eb/colab/chr8/i_2048_w_1/sequences.bed | sort | uniq -c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0b3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
